File: src/main/python/WordSubsets_916.py
from typing import List


class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        b = [0] * 26
        for word in words2:
            temp = [0]*26
            for c in word:
                temp[ord(c) - ord('a')] += 1
            for i in range(26):
                b[i] = max(b[i], temp[i])
        res = []
        # print(b)
        for word in words1:
            a = [0] * 26
            for c in word:
                a[ord(c) - ord('a')] += 1
            flag = True
            # print(a)
            for i in range(26):
                if b[i] > 0 and a[i] < b[i]:
                    flag = False
                    break
            if flag:
                res.append(word)
        return res

File: src/main/scala/WordSubsets_916.scala
object Solution {

  def frequency(word: String): Array[Int] ={
    word.foldLeft(Array.fill(26)(0)){
      (freq, c) => freq.updated(c - 'a', freq(c - 'a') + 1)
    }
  }

  def wordSubsets(words1: Array[String], words2: Array[String]): List[String] = {
    val words2Freq = words2.map(frequency).reduce {
      (freq1, freq2) => freq1.indices.map(i => math.max(freq1(i), freq2(i))).toArray
    }

    words1.filter { word =>
      val freq = frequency(word)
      words2Freq.indices.forall(i => freq(i) >= words2Freq(i))
    }.toList
  }
}